python/plugins/processing/algs/qgis/RasterCalculator.py

```diff
@@ -213,3 +225,69 @@ def processBeforeAddingToModeler(self, algorithm, model):
                         values.append(ValueFromOutput(alg.modeler_name, out.name))
 
         algorithm.params[self.LAYERS] = values
+
+    def mappedNameToLayer(self, lyr, expression, layersDict, context):
+        '''Try to identify if a real layer is mapped in the expression with a symbolic name.'''
+
+        nameToMap = lyr.name()
+
+        # get last scope of the expressionContext because should be that related
+        # with mapped variables
+        # The scope name should be "algorithm_inputs"
+        expContextLastScope = context.expressionContext().lastScope()
+
+        # check for layers directly added in the expression
+        if (nameToMap + "@") in expression:
+            layersDict[nameToMap] = lyr
+
+        # check for the layers that are mapped as input in a model
+        #  to do this check in the latest scope all passed variables
+        # to look for a variable that is a layer or a string filename ç
+        # to a layer
+        varId = None
+        varDescription = None
+
+        for varName in expContextLastScope.variableNames():
+
+            layer = expContextLastScope.variable(varName)
+
+            if not isinstance(layer, str) and not isinstance(layer, QgsMapLayer):
+                continue
+
+            if isinstance(layer, QgsMapLayer) and nameToMap not in layer.source():
+                continue
+
+            varId = varName
+            varDescription = expContextLastScope.description(varName)
+
+            # because there can be variable with None or "" description
+            # then skip them
+            if not varDescription:
+                continue
+
+            # check if it's description starts with Output as in:
+            #    Output 'Output' from algorithm 'calc1'
+            # as set in https://github.com/qgis/QGIS/blob/master/src/core/processing/models/qgsprocessingmodelalgorithm.cpp#L516
+            # but var in expression is called simply
+            #    'Output' from algorithm 'calc1'
+            elements = varDescription.split(" ")
+            if len(elements) > 1 and elements[0] == "Output":
+                # remove heading "Output " string
+                varDescription = varDescription[7:]
+
+            # check if cleaned varDescription is present in the expression
+            # if not skip it
+            if (varDescription + "@") not in expression:
+                continue
+
+            # !!!found!!! => substitute in expression
+            # and add in the list of layers that will be passed to raster calculator
```

Comment:
What about adding `if (layer == lyr.source()) or (layer == lyr):` here? The first test matches algorithm outputs, the second matches project layers. Tested and works with *Model 1* (this is to resolve my comment below at L291).

Reply:
This could also be inverted and added to your checks in L254-L258

    if (layer != lyr.source()) and (layer != lyr):
        continue

Reply:
layer == lyr.source() is already tested  in
https://github.com/luipir/QGIS/blob/8a05d6a3e43c5d17220a6f18cc5a41b63ba1b7d6/python/plugins/processing/algs/qgis/RasterCalculator.py#L257